### PR TITLE
525 output mappings

### LINKED ISSUE #525: Move output mappings under its own sub tree eg ${{steps.ws.outputs.myvar}}
* [ ] move step level vars under 'vars' 
* [ ] rename flow level config to 'vars' (do this deeply, var names and method names)
* [ ] add flow level config with enabled as the real config ie dataflow.config.enabled

### PR DESCRIPTION
**Pull request by jaypha**
Resolves #525 

Moved outputs to 'vars'.
Replaced dataflows.config with dataflows.vars.
Added config to dataflows for other settings.

### CODE REVIEW

File: db/upgrade.php [hunk 1]
```diff
@@ -141,5 +141,34 @@ function xmldb_tool_dataflows_upgrade($oldversion) {
...
+    if ($oldversion < 2022081000) {
+        // Changing type of field config on table tool_dataflows_steps to text.
+        $table = new xmldb_table('tool_dataflows');
+
+        $field = new xmldb_field('config', XMLDB_TYPE_TEXT, null, null, null, null, null, 'concurrencyenabled');
+        if ($dbman->field_exists($table, $field)) {
+            $dbman->rename_field($table, $field, 'vars');
+        }
+
+        $field = new xmldb_field('confighash', XMLDB_TYPE_TEXT, null, null, null, null, null, 'usermodified');
```
Comment (brendanheywood) [inline db/upgrade.php:153]: confighash should not have been caught up these changes, confighash represents the entirety of the config of the entire flow, steps, flow config, vars all of its. If anything changed at all then the "config" in the looser definition of the word has changed and so the hash should change

**Comment by jaypha** [comment]
confighash should not have changed

File: classes/dataflow.php [hunk 2]
```diff
@@ -746,49 +759,49 @@ public function get_export_data() {
...
             Yaml::DUMP_MULTI_LINE_LITERAL_BLOCK
         );
     }
 
     /**
-     * Saves the current config to the tool_dataflows_versions table
+     * Saves the current vars to the tool_dataflows_versions table
      */
-    public function save_config_version() {
+    public function save_vars_version() {
```
Comment (brendanheywood) [inline classes/dataflow.php:786]: this all looks like it should not be touched too

File: db/upgrade.php (hunk 1 above)
Comment (jaypha) [reply db/upgrade.php]: done

File: classes/dataflow.php (hunk 2 above)
Comment (jaypha) [reply classes/dataflow.php]: done

File: classes/dataflow.php [hunk 3]
```diff
@@ -773,21 +786,21 @@ public function set_var($name, $value) {
...
             $newconfighash = sha1($configyaml);
 
-            // Inserts the stored config if it is a new config.
+            // Inserts the stored vars if it is a new vars.
             if (!$DB->record_exists('tool_dataflows_versions', ['dataflowid' => $this->id, 'confighash' => $newconfighash])) {
                 $DB->insert_record('tool_dataflows_versions',
                     (object) ['dataflowid' => $this->id, 'confighash' => $newconfighash, 'configyaml' => $configyaml]);
             }
-            // Set the config hash in dataflows table.
+            // Set the vars hash in dataflows table.
```
Comment (brendanheywood) [inline classes/dataflow.php:803]: this comment can go?

File: classes/dataflow.php [hunk 4]
```diff
@@ -773,21 +786,21 @@ public function set_var($name, $value) {
...
-                (array) $config,
+                (array) $vars,
                 helper::YAML_DUMP_INLINE_LEVEL,
                 helper::YAML_DUMP_INDENT_LEVEL,
                 Yaml::DUMP_MULTI_LINE_LITERAL_BLOCK
             );
             $newconfighash = sha1($configyaml);
 
-            // Inserts the stored config if it is a new config.
+            // Inserts the stored vars if it is a new vars.
```
Comment (brendanheywood) [inline classes/dataflow.php:798]: I think this comment can go too

Comment (brendanheywood) [reply classes/dataflow.php:798]: the change I mean

File: db/install.xml [hunk 5]
```diff
@@ -10,7 +10,7 @@
         <FIELD NAME="name" TYPE="char" LENGTH="200" NOTNULL="true" SEQUENCE="false" COMMENT="The name of this dataflow"/>
         <FIELD NAME="enabled" TYPE="int" LENGTH="1" NOTNULL="true" DEFAULT="0" SEQUENCE="false"/>
         <FIELD NAME="concurrencyenabled" TYPE="int" LENGTH="1" NOTNULL="true" DEFAULT="0" SEQUENCE="false"/>
-        <FIELD NAME="config" TYPE="text" NOTNULL="false" SEQUENCE="false"/>
+        <FIELD NAME="vars" TYPE="text" NOTNULL="false" SEQUENCE="false" COMMENT="YAML encoded varaibles"/>
```
Comment (brendanheywood) [inline db/install.xml:13]: typo in spelling

File: edit.php [hunk 6]
```diff
@@ -84,8 +84,8 @@
             // Or, do the following if you don't want capability checks (discouraged).
             $persistent->from_record($data);
 
-            // Unset config hash if config has changed, so it will be resaved on next run.
-            if ($currentconfig !== $persistent->get('config')) {
+            // Unset config hash if vars has changed, so it will be resaved on next run.
+            if ($currentvars !== $persistent->get('vars')) {
```
Comment (brendanheywood) [inline edit.php:88]: I think this was not quite right in the original code, if *anything* changes in the configuration of the flow then the config hash should change, so changes to enable or concurrency or the vars. I think just make it easy and if you save the record then we should reset the hash. If it regenerates and comes back with the same hash then no big deal

Comment (jaypha) [reply edit.php]: fixed

File: db/install.xml (hunk 5 above)
Comment (jaypha) [reply db/install.xml]: done

File: classes/dataflow.php (hunk 4 above)
Comment (jaypha) [reply classes/dataflow.php]: done

File: classes/dataflow.php (hunk 3 above)
Comment (jaypha) [reply classes/dataflow.php]: done

**Comment by brendanheywood** [comment]
+1 from me, the most important thing is the upgrade script, if there are other little bugs its ok and we can fix them in followup issues

**Pull request merged**